### PR TITLE
[SYCL] Transition from ONEAPI/INTEL to ext::oneapi/intel namespaces

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
To be conformant with SYCL 2020, this patch deprecates `ONEAPI` and
`INTEL` namespaces without breaking changes in ABI, and introduces
new `ext::oneapi` and `ext::intel` namespaces.

This patch does:

1. **RT part:** deprecate `ONEAPI` and `INTEL` namespaces, introduce the 
replacement: `ext::oneapi` and `ext::intel` namespaces. Copy
include/CL/sycl/ONEAPI to sycl/ext/oneapi and include/CL/sycl/INTEL to
sycl/ext/intel as is. Add deprecation warnings to include/CL/sycl/ONEAPI
and include/CL/sycl/INTEL headers.
2. **FE and Tools part:** replace `ONEAPI`/`INTEL` namespaces with 
`ext::oneapi/intel`, update mangling in tests. Modify `Util::isSycl...` funcs
to accept both deprecated and new namespaces for compatibility reasons.

### CODE REVIEW

**Review by Pennycook — CHANGES_REQUESTED** [review]
According to [SYCL 2020](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_guidelines_for_portable_extensions), the new namespace should be `sycl::ext::oneapi`, not `sycl::oneapi`:

> The namespace `sycl::ext::<vendorstring>` is reserved for use by extensions. For example, the Acme vendor could define extended types and free functions in the namespace `sycl::ext::acme`, and this would guarantee that they will not collide with definitions in other vendors' extensions or with future versions of the core SYCL specification.

**Comment by dm-vodopyanov** [comment]
> According to SYCL 2020, the new namespace should be sycl::ext::oneapi, not sycl::oneapi

Thanks, fixed.

**Review by bader — DISMISSED** [review]
.github/CODEOWNERS changes look good to me.

File: llvm/include/llvm/SYCLLowerIR/LowerESIMD.h
Comment (kbobrovs) [inline llvm/include/llvm/SYCLLowerIR/LowerESIMD.h:13]: Since you are fixing this anyway, could you please change 'sycl::ext::intel::gpu' to 'sycl::ext::intel::experimental::esimd' here and below?

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp
Comment (kbobrovs) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:129]: unrelated change - please remove

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/esimd_global.ll:9]: no need to change here. This is incorrect name anyway (see correct one in my comment above). The type names are meaningless in LLVM IR, just serve as symbolic IDs.

**Review by MrSidims — APPROVED** [review]
FPGA changes are LGTM
+ @GarveyJoe for awareness

File: llvm/include/llvm/SYCLLowerIR/LowerESIMD.h
Comment (dm-vodopyanov) [reply llvm/include/llvm/SYCLLowerIR/LowerESIMD.h]: Changed in https://github.com/intel/llvm/pull/4014/commits/5f8dfa54758be4bc71d9169091a22e45d449a961

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp
Comment (dm-vodopyanov) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp]: Removed in https://github.com/intel/llvm/pull/4014/commits/5f8dfa54758be4bc71d9169091a22e45d449a961

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (dm-vodopyanov) [reply llvm/test/SYCLLowerIR/esimd_global.ll:9]: Mb we can introduce the replacement from INTEL to ext::intel in this patch, and after that these functions w incorrect names will be replaced with the correct ones as separate PR?

File: sycl/test/warnings/sycl_2020_deprecations.cpp
Comment (bader) [inline sycl/test/warnings/sycl_2020_deprecations.cpp:7]: Is this change required?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:5072]: Should we add a warning for using deprecated  namespace @AaronBallman @premanandrao ?

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/Inputs/sycl.hpp:324]: Unrelated clang-format change?

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/Inputs/sycl.hpp:339]: Unrelated clang-format change?

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/Inputs/sycl.hpp:447]: Here too?

**Review by Pennycook — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:5072]: Yes; as a rule of thumb, when we deprecate something we should alert users when they're using it. Otherwise the deprecation doesn't help us to remove anything (we'll just silently break their code, which we could do without bothering to have a deprecation period at all).

Comment (dm-vodopyanov) [reply clang/lib/Sema/SemaSYCL.cpp:5072]: We have deprecation warnings. See e.g., https://github.com/intel/llvm/pull/4014/files#diff-75fc8ffcd0922ca67040695dbb6811aacc406812b55436b29f171cbac2e44c92R283

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (dm-vodopyanov) [reply clang/test/CodeGenSYCL/Inputs/sycl.hpp]: Fixed

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (dm-vodopyanov) [reply clang/test/CodeGenSYCL/Inputs/sycl.hpp]: Fixed

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp
Comment (dm-vodopyanov) [reply clang/test/CodeGenSYCL/Inputs/sycl.hpp]: Fixed

File: sycl/test/warnings/sycl_2020_deprecations.cpp
Comment (dm-vodopyanov) [reply sycl/test/warnings/sycl_2020_deprecations.cpp]: Not anymore, reverted.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:5072]: Ok. Thank you for confirmation

**Review by elizabethandrews — DISMISSED** [review]

**Review by bader — COMMENTED** [review]
.github/CODEOWNERS changes look good to me.

**Review by bader — DISMISSED** [review]

**Review by Pennycook — DISMISSED** [review]

**Review by MrSidims — DISMISSED** [review]

**Review by smaslov-intel — DISMISSED** [review]

File: llvm-spirv/test/transcoding/intel_fpga_lsu_optimized.ll
Comment (AlexeySachkov) [inline llvm-spirv/test/transcoding/intel_fpga_lsu_optimized.ll:5]: I think this change can be omitted completely, because it is quite useless if not upstreamed

File: llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll:42]: I also think that those changes are not strictly necessary and from test readability point of view I would prefer shorter made up names rather than ones which are strictly the same as generated by FE from Headers

File: sycl/include/CL/sycl/properties/function_pointer.hpp
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/properties/function_pointer.hpp:1]: This seems like an extra copy, the same comment applies to a few more files in `sycl/include/CL/sycl/properties` directory

**Review by AlexeySachkov — DISMISSED** [review]
8 files which I cover as a code owner LGTM. However, I think that some changes are unnecessary.

I also noticed a few seemingly extra copies of some extensions header files

**Comment by gmlueck** [comment]
It looks like the documentation for all these extensions did not get updated.  Shouldn't that be part of this PR also?  For example "sycl/doc/extensions/IntelFPGA/FPGASelector.md" still lists the old header file path and the old namespace name.

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (dm-vodopyanov) [reply llvm/test/SYCLLowerIR/esimd_global.ll]: Fixed

File: llvm-spirv/test/transcoding/intel_fpga_lsu_optimized.ll
Comment (dm-vodopyanov) [reply llvm-spirv/test/transcoding/intel_fpga_lsu_optimized.ll]: Removed

File: llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll
Comment (dm-vodopyanov) [reply llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll]: Reverted

File: sycl/include/CL/sycl/properties/function_pointer.hpp
Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/properties/function_pointer.hpp]: Good catch! Removed

**Comment by dm-vodopyanov** [comment]
> It looks like the documentation for all these extensions did not get updated. Shouldn't that be part of this PR also? For example "sycl/doc/extensions/IntelFPGA/FPGASelector.md" still lists the old header file path and the old namespace name.

Yes, it's a good place to update them now. Updated in 2611c49 and 8a6c9a2

File: llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll
Comment (AlexeySachkov) [reply llvm/test/tools/sycl-post-link/spec-constants/multiple-composite-usages.ll]: Note: there are 4 more spec constant tests + a bunch of ESIMD tests which also can be left as-is, I guess (but better ask @dendibakh or @kbobrovs)

File: sycl/ReleaseNotes.md
Comment (AlexeySachkov) [inline sycl/ReleaseNotes.md:13]: What is the point of editing old release notes?

I think that deprecation notice should appear in the new section for the next release instead of being hidden in already published notes.

Tagging @romanovvlad and @bader for their feedback

**Review by AlexeySachkov — DISMISSED** [review]
Files owned by me LGTM

File: sycl/ReleaseNotes.md
Comment (dm-vodopyanov) [reply sycl/ReleaseNotes.md:13]: Yes, I also revised this - this is not a good idea because we add changes from newer release to the older release, the only pros is that if customer uses newer version of the compiler where we removed sycl::ONEAPI and wanted to see what features can be used from older releases, they will be confused after finding out that e.g., there is no sycl::ONEAPI::has_known_identity in the product.
I will revert this.

Comment (dm-vodopyanov) [reply sycl/ReleaseNotes.md]: Done

**Comment by dm-vodopyanov** [comment]
@AlexeySachkov @kbobrovs @elizabethandrews @bader @Pennycook @MrSidims @smaslov-intel , can you please approve this one last time? The checks for penultimate commit passed successfully, the last commit has the only changes in docs, so checks should be passed for it as well.

**Review by AlexeySachkov — DISMISSED** [review]

**Review by MrSidims — DISMISSED** [review]
FPGA changes LGTM

**Review by elizabethandrews — DISMISSED** [review]
FE changes LGTM

**Review by Pennycook — DISMISSED** [review]

**Review by smaslov-intel — APPROVED** [review]

File: sycl/include/sycl/ext/intel/esimd.hpp
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd.hpp:1]: why this file is added?
The vanilla location for esimd.hpp is `sycl/ext/intel/experimental/esimd.hpp`, why add one more redirection?

File: sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h:1]: here and in all other esimd files: correct location is `sycl/ext/intel/experimental/esimd/emu/detail`

File: sycl/include/CL/sycl/INTEL/esimd/detail/emu/esimdcpu_device_interface.hpp
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/detail/emu/esimdcpu_device_interface.hpp:26]: here and in similar places:
```suggestion
    "sycl/ext/intel/experimental/esimd/emu/detail/esimdcpu_device_interface.hpp instead")
```
see also comment about correct esimd files location below

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/esimd_global.ll:9]: Here and in all other similar places. Correct namespace for public esimd apis is `sycl::ext::intel::experimental::esimd`. E.g.
`sycl::ext::intel::experimental::esimd::simd`. My suggestion is either not to touch the .ll files at all, or change to "namespace-neutral" form like `%"class.esimd::simd_IiLi16"`. Note the `IiLi16` stem, which is needed to unique from other simd class instantiations within LLVMIR.

**Review by kbobrovs — CHANGES_REQUESTED** [review]

**Comment by kbsmith-intel** [comment]
What has been done to make sure sycl library built with these changes is backward compatible (has every entry point and interface is the same) as the sycl library before these changes?

**Comment by dm-vodopyanov** [comment]
> What has been done to make sure sycl library built with these changes is backward compatible (has every entry point and interface is the same) as the sycl library before these changes?

The idea is the following (e.g., for `ONEAPI::`): 
1. All content from the original `ONEAPI::` namespace moves to `ext::oneapi::`, after that `[[deprecated]] ONEAPI::` namespace is created which contains `using namespace ext::oneapi;`
2. To keep all exported symbols from the library which contain `6ONEAPI` substr (there are few of them), wrappers over exported `ext::oneapi::` data structures were created in `ONEAPI::` namespace, e.g., see the example for ext::oneapi::filter_selector: [here](https://github.com/intel/llvm/pull/4014/files#diff-a42d3e2c9a3ff05d9355b4303bb56922775401fb2ba547d64ae1093d961a4586R46) and [here](https://github.com/intel/llvm/pull/4014/files#diff-5f1bc6220fa47d24b2edf62bd05b940c9506efe02b77f54ecb21ce636da446abR275).

After this change exported data structures for `ONEAPI::` and `INTEL::` namespaces will be compiled twice but they are really few: 
* online_compiler class (in `INTEL::` and `ext::intel::`), 
* filter_selector class (in `ONEAPI::` and `ext::oneapi::`) 
* free functions reduGetMaxWGSize and reduComputeWGSize (in `ONEAPI::detail::` and `ext::oneapi::detail::`)

No more exported functions.

File: sycl/include/sycl/ext/intel/esimd.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/intel/esimd.hpp:1]: @kbobrovs , from the description:
```
Copy
include/CL/sycl/ONEAPI to sycl/ext/oneapi and include/CL/sycl/INTEL to
sycl/ext/intel as is.
```

File: sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h:1]: Same, it was done like that:
```
Copy
include/CL/sycl/ONEAPI to sycl/ext/oneapi and include/CL/sycl/INTEL to
sycl/ext/intel as is.
```
What do you propose for esimd? Don't touch esimd code in include/CL/sycl/INTEL (except marking namespace as deprecated)?

**Review by bader — APPROVED** [review]

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (dm-vodopyanov) [reply llvm/test/SYCLLowerIR/esimd_global.ll:9]: @kbobrovs are you ok to address all your comments in the separate PR due to upcoming internal feature freeze? The purpose of this feature was to make a transition from INTEL/ONEAPI namespaces to ext::intel/oneapi w/o changing any logic.

**Review by kbsmith-intel — DISMISSED** [review]
LGTM

File: sycl/include/sycl/ext/intel/esimd.hpp
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/esimd.hpp:1]: anyway, sycl/include/sycl/ext/intel/esimd.hpp should be removed. include/CL/sycl/INTEL/esimd.h header redirection is already in-place, so there is no need to add any more files.

File: sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h:1]: INTEL/esimd.h is already OK (maybe the `#include elementary` should be moved to the redirected to header.
```cpp
#pragma once

#include <CL/sycl/detail/defines_elementary.hpp>

__SYCL_WARNING("CL/sycl/INTEL/esimd.hpp usage is deprecated, include "
               "sycl/ext/intel/experimental/esimd.hpp instead")

#include <sycl/ext/intel/experimental/esimd.hpp>
```

`INTEL/esimd/detail/emu` should be moved to `sycl/ext/intel/experimental/esimd/emu/detail` for now. We need to review namespace in esimdcpu_device_interface.h and move further if needed. @dongkyunahn-intel - FYI.

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (kbobrovs) [reply llvm/test/SYCLLowerIR/esimd_global.ll:9]: @dm-vodopyanov - you mean dropping ESIMD changes from  this PR, and creating a new one? Yes, I'm OK, if so.

File: sycl/include/sycl/ext/intel/esimd.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/intel/esimd.hpp]: Done

File: sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h]: Done

File: sycl/include/CL/sycl/INTEL/esimd/detail/emu/esimdcpu_device_interface.hpp
Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/INTEL/esimd/detail/emu/esimdcpu_device_interface.hpp]: Done

File: llvm/test/SYCLLowerIR/esimd_global.ll
Comment (dm-vodopyanov) [reply llvm/test/SYCLLowerIR/esimd_global.ll]: No, I meant to commit PR as is to have time to cherry-pick to internal repo, and apply comments as separate patch and also cherry-pick it but after feature freeze.

Anyway, I already applied comments.

Comment (dm-vodopyanov) [reply llvm/test/SYCLLowerIR/esimd_global.ll]: Reverted such changes in *.ll files

**Review by kbobrovs — APPROVED** [review]

**Comment by dm-vodopyanov** [comment]
Waiting for CI to complete. @bader, @vladimirlaz, @romanovvlad feel free to merge the patch if all checks pass as all required approvals were received.

File: sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h]: I already have changes for moving files in the `INTEL/esimd/detail/emu` in my PR for ESIMD_CPU support - https://github.com/intel/llvm/pull/4011/commits/7d78e713b496390fdca90755487430bf1cbcd374.

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/esimd/detail/emu/esimd_emu_functions_v1.h]: I'll revert the above change in my PR when this PR is merged in the `origin/sycl` branch.

**Comment by dm-vodopyanov** [comment]
@bader I'm looking into compile failure on post-commit with SharedLibs.

**Pull request merged**
